analysis/src/TypeUtils.ml

```diff
@@ -148,18 +148,90 @@ let rec extractType ~env ~package (t : Types.type_expr) =
-    | _ -> (env, t))
-  | _ -> (env, t)
+let findReturnTypeOfFunctionAtLoc loc ~(env : QueryEnv.t) ~full ~debug =
+  match References.getLocItem ~full ~pos:(loc |> Loc.end_) ~debug with
+  | Some {locType = Typed (_, typExpr, _)} -> (
+    match extractFunctionType ~env ~package:full.package typExpr with
+    | args, tRet when args <> [] -> Some tRet
+    | _ -> None)
+  | _ -> None
+
+type builtinType =
+  | Array
+  | Option
+  | String
+  | Int
+  | Float
+  | Promise
+  | List
+  | Result
+  | Lazy
+  | Char
+
+type pipeCompletionType =
+  | Builtin of builtinType * Types.type_expr
+  | TypExpr of Types.type_expr
+
+let getBuiltinFromTypePath path =
+  match path with
+  | Path.Pident id when Ident.name id = "array" -> Some Array
+  | Path.Pident id when Ident.name id = "option" -> Some Option
+  | Path.Pident id when Ident.name id = "string" -> Some String
+  | Path.Pident id when Ident.name id = "int" -> Some Int
+  | Path.Pident id when Ident.name id = "float" -> Some Float
+  | Path.Pident id when Ident.name id = "promise" -> Some Promise
+  | Path.Pident id when Ident.name id = "list" -> Some List
+  | Path.Pident id when Ident.name id = "result" -> Some Result
+  | Path.Pident id when Ident.name id = "lazy_t" -> Some Lazy
+  | Path.Pident id when Ident.name id = "char" -> Some Char
+  | Pdot (Pident id, "result", _) when Ident.name id = "Pervasives" ->
+    Some Result
+  | _ -> None
+
+let rec resolveTypeForPipeCompletion ~env ~package ~lhsLoc ~full
+    (t : Types.type_expr) =
+  let builtin =
+    match t.desc with
+    | Tconstr (path, _typeArgs, _)
+    | Tlink {desc = Tconstr (path, _typeArgs, _)}
+    | Tsubst {desc = Tconstr (path, _typeArgs, _)}
+    | Tpoly ({desc = Tconstr (path, _typeArgs, _)}, []) ->
+      path |> getBuiltinFromTypePath
+    | _ -> None
+  in
+  match builtin with
+  | Some builtin -> (env, Builtin (builtin, t))
+  | None -> (
+    (* If the type we're completing on is a type parameter, we won't be able to
+       do completion unless we know what that type parameter is compiled as.
+       This attempts to look up the compiled type for that type parameter by
+       looking for compiled information at the loc of that expression. *)
+    let typFromLoc =
+      match t with
+      | {Types.desc = Tvar _} -> (
+        match findReturnTypeOfFunctionAtLoc lhsLoc ~env ~full ~debug:false with
+        | None -> None
+        | Some typFromLoc -> Some typFromLoc)
+      | _ -> None
+    in
+    match typFromLoc with
+    | Some typFromLoc ->
+      typFromLoc |> resolveTypeForPipeCompletion ~lhsLoc ~env ~package ~full
+    | None ->
+      let rec digToRelevantType ~env ~package (t : Types.type_expr) =
+        match t.desc with
+        | Tlink t1 | Tsubst t1 | Tpoly (t1, []) ->
+          digToRelevantType ~env ~package t1
+        (* Don't descend into types named "t". Type t is a convention in the ReScript ecosystem. *)
+        | Tconstr (path, _, _) when path |> Path.last = "t" -> (env, TypExpr t)
```

Comment:
What happens in practice if this case is omitted?
I was thinking of the different cases. E.g. when it's an alias to `array`, whether one wants to complete for `array`. And when it's a record definition, then one wants to treat is as the source of truth.
Just thinking about here, about what the convention could look like.

Reply:
Yeah, I added this case because it broke 2 tests in `Completion.res`, and looking at them I felt like the old behavior was better.

Yeah, I agree. I don't feel like there's a clear and good answer here, but I do lean a bit towards the current behavior, as in `type t` usually means the main type, and if it's in a module there's often functions in the module to work on it. But I agree it's not 100%. 

One idea could be to complete for the module with the type `t`, _and_ the aliased value. Unsure if that'd be messy to implement, but it might be good behavior. What do you think?

Merging since it doesn't change current behavior.